language/ast.py

```diff
@@ -0,0 +1,678 @@
+from langkit.dsl import ASTNode, Field, abstract
+
+
+@abstract
+class RFLXNode(ASTNode):
+    """
+    Root node class for the RecordFlux language
+    """
+
+
+@abstract
+class AbstractID(RFLXNode):
+    """
+    Base class for identifiers
+    """
+
+
+class NullID(AbstractID):
+    """
+    "null" identifier
+    """
+
+
+class UnqualifiedID(AbstractID):
+    """
+    Simple, unqualified identifiers, i.e. identifiers without a package part (e.g. "Foo")
+    """
+
+    token_node = True
+
+
+class ID(AbstractID):
+    """
+    Qualified identifiers which may optionally have a package part (e.g. "Pkg::Foo", "Foo")
+    """
+
+    package = Field(type=UnqualifiedID)
+    name = Field(type=UnqualifiedID)
+
+
+@abstract
+class Declaration(RFLXNode):
+    """
+    Base class for declarations (types, refinements, sessions)
+    """
+
+
+@abstract
+class TypeDef(RFLXNode):
+    """
+    Base class for type definitions (integers, abstract messages, type derivations, arrays,
+    enumerations)
+    """
+
+
+class TypeDecl(Declaration):
+    """
+    Type declaration (type Foo is ...)
+    """
+
+    identifier = Field(type=UnqualifiedID)
+    definition = Field(type=TypeDef)
+
+
+@abstract
+class Expr(RFLXNode):
+    """
+    Base class for expressions
+    """
+
+
+class Op(RFLXNode):
+    """
+    Operators for expressions
+    """
+
+    enum_node = True
+    alternatives = [
+        "pow",
+        "mul",
+        "div",
+        "add",
+        "sub",
+        "mod",
+        "eq",
+        "neq",
+        "le",
+        "lt",
+        "gt",
+        "ge",
+        "and",
+        "or",
+        "in",
+        "notin",
```

Comment:
It seems that `not` is missing in this list. Also does it make sense to have `notin` instead of `not` `in`?

Reply:
This is just the definition of the *binary* operators available in the AST (this is not what matches the input text). I clarify that those are *binary* operators in dce7445.